### PR TITLE
fix(vega): validate raw SQL resource placeholders

### LINKED ISSUE #400: [Bug][Vega] Resource 占位符藏在 SQL 注释中仍可通过绑定校验
## 问题描述

Vega Raw Query 文档要求 SQL 在 `FROM` / `JOIN` 中使用 `{{resource-id}}` 占位符，由服务端绑定 Resource 对应的物理表。

当前实现会先用正则从原始 SQL 文本提取 Resource ID，但没有排除 SQL 注释。调用方可以在正文直接写物理表名，只把 `{{resource-id}}` 放在块注释中；只要物理表与该 Resource 的 `source_identifier` 相同，查询仍会通过。

这是 #275 只读门禁上线后的接口约束缺陷。它目前不能绕过只读限制，也不能访问其他未绑定表，但实现与文档不一致，并为后续 SQL 扩展留下权限校验风险。

## 环境

- 平台：`https://14.103.77.23`
- 日期：2026-07-23
- Vega backend：`0.1.1-main.20260723103308.sha0404737`
- SDK：`bkn-sdk` main 源码构建
- Catalog：`test`（`d9etnengant5l9p8jq90`）
- Resource：`d9etnv7gant5l9p8m8dg`
- `source_identifier`：`test.t_5000`

### 1. 正常查询

```bash
node dist/cli.js -k --json vega sql \
  --input-dialect mysql \
  --query "SELECT * FROM {{d9etnv7gant5l9p8m8dg}} LIMIT 1"
```

结果：成功返回 `columns / entries / paging`。

### 2. 把 Resource 占位符藏在注释中

```bash
node dist/cli.js -k --json vega sql \
  --input-dialect mysql \
  --query "SELECT * FROM test.t_5000 /* {{d9etnv7gant5l9p8m8dg}} */ LIMIT 1"
```

实际结果：仍成功返回 `columns / entries / paging`，问题复现。

### 3. 跨表对照

使用上述 Resource ID 查询其他物理表时，返回 HTTP 400：

```sql
SELECT * FROM test.t_4999 /* {{d9etnv7gant5l9p8m8dg}} */ LIMIT 1
```

说明物理表白名单仍有效，但占位符使用位置未被正确校验。

## 实际结果

Resource ID 只存在于 `/* ... */` 注释中，没有在 `FROM` 表引用位置参与绑定，查询仍被放行。

## 期望结果

第二条 SQL 应在连接器执行前返回 HTTP 400，例如：

```text
Resource placeholder must be used as a FROM/JOIN table reference
```

以下位置中的 `{{resource-id}}` 均不应作为 Resource 绑定依据：

- `-- {{resource-id}}` 行注释
- `/* {{resource-id}} */` 块注释
- `'{{resource-id}}'` 字符串字面量

## 原因分析

`extractResourceIDs` 当前通过正则扫描原始 SQL：

```go
regexp.MustCompile(`\{\{\.?([a-z0-9][a-z0-9_-]{0,39})\}\}`)
```

正则无法区分占位符位于表引用、注释还是字符串。后续表白名单校验只确认解析出的物理表属于已绑定 Resource，因此同表场景会放行。

相关代码：

- `adp/vega/vega-backend/server/logics/query/raw_query_service.go`
- `adp/vega/vega-backend/server/logics/query/querypolicy/sqlglot_adapter.go`

## 修复建议

1. 在提取 Resource ID 前通过词法/AST 预处理排除注释和字符串字面量。
2. 更稳妥的方案是将 Resource 占位符转换为安全临时标识后解析 AST，确认它确实出现在 `FROM` / `JOIN` 的 `Table` 节点。
3. 替换占位符后继续保留现有只读 AST 校验和物理表白名单校验。
4. 增加块注释、行注释、字符串、CTE、子查询、JOIN 和大小写混淆回归测试。

## 验收标准

- `FROM {{resource-id}}` 正常查询成功。
- 占位符仅位于块注释、行注释或字符串时返回 HTTP 400。
- A Resource 不能查询 B 物理表。
- DDL/DML、多语句和锁定读仍被只读策略拒绝。
- 错误响应和日志不泄露 SQL 全文、物理表名或连接信息。

关联：#275

### PR DESCRIPTION
**Pull request by kenlee1988**
## What Changed

Brief description of changes:

- [x] Validate SQL Resource placeholders only when they appear in `FROM`/`JOIN` table references.
- [x] Preserve placeholders in comments and string literals during binding.
- [x] Add focused regression coverage for block comments, line comments, string literals, and valid joins.
- [ ] Docs/doc 1 — Not applicable; no public API contract change.

---

## Why

- Related Issue: Closes #400
- Related Feature: Not applicable
- Background: Raw SQL previously accepted a Resource placeholder found only in a comment or string literal when the directly named physical table matched that Resource.

---

## How

- Key implementation points: replace code-region placeholders with unique temporary identifiers, parse with SQLGlot, and collect only `Table` AST nodes; return 400 when no valid table-bound Resource exists.
- Breaking changes (API, config, database, etc.): Queries that use only direct physical table names or placeholders in comments/literals now return HTTP 400, consistent with the documented Resource-binding contract.

---

## Testing

- [x] Unit tests passed locally
- [ ] Integration tests passed locally — Not run; no external dependency is needed for this parser and service-layer change.
- [ ] Verified in test environment — Not run.

Test notes:

- `make test`
- `make lint`
- `go test ./logics/query ./logics/query/querypolicy -count=1`

---

## Risk & Rollback

- Potential risks: SQL dialect edge cases involving quoted identifiers or uncommon literal syntax may require additional compatibility cases.
- Rollback plan: Revert commit `a023e421`.

---

## Additional Notes

- The working-tree change in `server/config/vega-backend-config.yaml` is user-owned and is not included in this PR.

### CODE REVIEW

File: adp/vega/vega-backend/server/logics/query/querypolicy/sqlglot_adapter.go [hunk 1]
```diff
@@ -114,6 +114,117 @@ type validationResult struct {
...
+            quote = sql[index]
+            end = index + 1
+            while end < len(sql):
+                if sql[end] == quote:
+                    if end + 1 < len(sql) and sql[end + 1] == quote:
+                        end += 2
+                        continue
+                    end += 1
+                    break
+                if sql[end] == "\\\\" and quote == "'" and end + 1 < len(sql):
```
[the remark this answers is not in the document]
Comment (kenlee1988) [reply adp/vega/vega-backend/server/logics/query/querypolicy/sqlglot_adapter.go]: 已在 dd524425 修复。Go 与 Python 现仅在 MySQL 方言将反斜杠视为单引号转义，词法行为一致；新增 MySQL 反斜杠转义字符串的回归用例。资源提取也已纳入 QueryPolicy Adapter，因此单元测试可注入 mock，不再依赖 CI 环境安装 SQLGlot。已本地通过 `make test-cover COVERAGE_THRESHOLD=10` 和 `make lint`。

File: adp/vega/vega-backend/server/logics/query/raw_query_service_test.go [hunk 2]
```diff
@@ -200,12 +209,50 @@ func TestRawQueryServiceValidateRequestNewContract(t *testing.T) {
...
+	t.Cleanup(func() { rawQueryPolicy = previousPolicy })
+
+	ids, err := (&rawQueryService{}).extractResourceIDs(context.Background(), "SELECT * FROM {{orders-2026}} JOIN {{.customer_data}} ON true", "postgres")
 
 	require.NoError(t, err)
 	assert.Equal(t, []string{"orders-2026", "customer_data"}, ids)
 }
 
+func TestPrepareSQLQueryRejectsResourcePlaceholderOutsideTableReference(t *testing.T) {
+	svc := &rawQueryService{}
```
[the remark this answers is not in the document]
Comment (kenlee1988) [reply adp/vega/vega-backend/server/logics/query/raw_query_service_test.go:223]: 已在 72c70c77 修复：测试改为注入返回空 Resource ID 的 Policy mock，并断言 `error_details` 为 `at least one resource_id is required for SQL queries`。因此它不再依赖 SQLGlot 运行时，且只会在目标“占位符未建立绑定”分支通过。

**Review by sh00tg0a1 — APPROVED** [review]
